### PR TITLE
Import Roster modal

### PR DESCRIPTION
**Pull request by cainlevy**
This is a MVP for importing a roster directly from the DOC client. I chose a modal, opened via a new `[Import]` button on the Incarcerated People page. Currently it is only available to super admins, although the backend permits all facility admins.

## Demo

https://user-images.githubusercontent.com/4449/153097757-469f92ce-1684-4e2d-ae04-f8ae73e7a701.mov

## Included

- schema validations
- data validations on the facilityName and groupName fields
- rudimentary error feedback

## Not Included

- a template for valid CSV files
- a call to action for automating uploads via our REST API
- parsing for REST API-compatible JSON data

### CODE REVIEW

File: src/components/ImportRoster/ImportModal.tsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+      data.facility.publicId,
+      ...(data.facility.system?.facilities.map((f) => f.publicId) || []),
+    ]);
+
+    const groupNames = uniq([
+      ...data.facility.groups.map((g) => g.name),
+      ...(data.facility.system?.facilities.flatMap((f) =>
+        f.groups.map((g) => g.name)
+      ) || []),
+    ]);
```
Comment (cainlevy) [inline src/components/ImportRoster/ImportModal.tsx:54]: grabbing all of the valid facility and group codes is a bit messy when systems are optional

File: src/components/ImportRoster/RosterDropzone.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    );
+  });
+};
+
+const RosterDropzone: React.FC<Props> = ({ onDrop, onError }) => {
+  const { getRootProps, getInputProps, isDragActive, fileRejections } =
+    useDropzone({
+      accept: "text/csv",
+      maxFiles: 1,
+      onDropAccepted: ([file]) => parseFileAsCSV(file).then(onDrop, onError),
```
Comment (cainlevy) [inline src/components/ImportRoster/RosterDropzone.tsx:51]: TODO: detect and parse as JSON

File: src/pages/InmatesPage.tsx [hunk 3]
```diff
@@ -97,6 +103,11 @@ const InmatesPage: React.FC = () => {
               onChange={(event) => setSearchQuery(event.target.value)}
             />
           </Input.Group>,
+          isSuperUser(session.user) ? (
+            <Button key="import" onClick={() => setShowImportModal(true)}>
+              Import
+            </Button>
+          ) : undefined,
```
Comment (cainlevy) [inline src/pages/InmatesPage.tsx:110]: we should try it and resolve some basic TODOs before releasing to facility admins.

File: src/components/ImportRoster/ImportModal.tsx [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  groupName: string;
+  dateOfBirth: Date;
+};
+
+type ValidatedResult = { inmate: InmateData; error: undefined };
+type ErroredResult = { inmate: undefined; error: Joi.ValidationError };
+
+const ImportModal: React.FC<Props> = ({ onClose }) => {
+  const [results, setResults] = useState<(ValidatedResult | ErroredResult)[]>();
+  const errors = results?.filter((i): i is ErroredResult => !!i.error) || [];
```
Comment (gabrielsaruhashi) [inline src/components/ImportRoster/ImportModal.tsx:31]: i never used this assertion before! i imagine this is for type narrowing so TS doesn't complain that error doesn't exist on ValidatedResult?

File: src/components/ImportRoster/ImportModal.tsx [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    if (!data) throw new Error("facility data has not loaded yet");
+
+    const facilityCodes = uniq([
+      data.facility.publicId,
+      ...(data.facility.system?.facilities.map((f) => f.publicId) || []),
+    ]);
+
+    const groupNames = uniq([
+      ...data.facility.groups.map((g) => g.name),
+      ...(data.facility.system?.facilities.flatMap((f) =>
```
Comment (gabrielsaruhashi) [inline src/components/ImportRoster/ImportModal.tsx:51]: if system exists the selected facility groups would be duped here right?

would it be more readable if we had

`data.facility.system ?    
     data.facility.system.facilities.flatMap((f) => f.groups.map((g) => g.name))) : 
     data.facility.groups.map((g) => g.name)`

File: src/components/ImportRoster/ImportModal.tsx [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        >
+          <ol className="max-h-52 overflow-y-scroll">
+            {errors.slice(0, 26).map(({ error }, idx) =>
+              idx < 25 ? (
+                <li key={idx} className="mb-4">
+                  {idx + 1}. {error!.details.map((d) => d.message).join(", ")}
+                  <pre>{JSON.stringify(error?._original)}</pre>
+                </li>
+              ) : (
+                <li key={idx}>& more</li>
```
Comment (gabrielsaruhashi) [inline src/components/ImportRoster/ImportModal.tsx:126]: should we display the total number of others errors?

i.e & {errors.length - 25} more

File: src/components/ImportRoster/ImportModal.tsx [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+              type="primary"
+              onClick={handleSubmit}
+              disabled={mutationLoading}
+            >
+              Submit
+            </Button>,
+          ]}
+        />
+      ) : loading ? (
+        <p>Loading facility data for validations ...</p>
```
Comment (gabrielsaruhashi) [inline src/components/ImportRoster/ImportModal.tsx:148]: nit: wanna add a spinner here?

File: src/components/ImportRoster/ImportModal.tsx [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+  };
+
+  return (
+    <Modal
+      title="Import Roster"
+      footer={null}
+      visible
+      onCancel={() => onClose()}
+    >
+      {errors.length > 0 ? (
```
Comment (gabrielsaruhashi) [inline src/components/ImportRoster/ImportModal.tsx:112]: should we move this to a sub-component soon so we don't have 3 nested ternary operators?

**Review by gabrielsaruhashi — APPROVED** [review]
alt

File: src/components/ImportRoster/ImportModal.tsx (hunk 4 above)
Comment (cainlevy) [reply src/components/ImportRoster/ImportModal.tsx:31]: yep!

https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

File: src/components/ImportRoster/ImportModal.tsx (hunk 5 above)
Comment (cainlevy) [reply src/components/ImportRoster/ImportModal.tsx:51]: hmm, yeah, i think that would mean i can drop the uniq

Comment (cainlevy) [reply src/components/ImportRoster/ImportModal.tsx:51]: oh this is just a lot simpler when i extract `const facilities` first.

File: src/components/ImportRoster/ImportModal.tsx (hunk 8 above)
Comment (cainlevy) [reply src/components/ImportRoster/ImportModal.tsx:36]: i extracted the components but left the ternary. in JSX it functions as an `if / elsif / else` statement.

**Pull request merged**